### PR TITLE
test(stream): add ci for split change mutation in source

### LINKED ISSUE #1285: TODO: support config change for assign split in SourceExecutor
SourceExecutor will rebuild a ConnectorSource every time receiving an `assign-split` mutation and then turn it into streaming when `next` is called later. 

- [x] impl a unified function to build all kinds of SplitReader in `rust/connector`
- [x] remove `CreateSourceExecutor` from `rust/batch`
- [x] introduce assign-split mutation in stream

Here we introduce a new assumption that assign-split mutation must happen before the first `next` called. Otherwise SourceExecutor will give empty StreamChunk or even panic.

### PR DESCRIPTION
**Pull request by tabVersion**
## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

add a test for config change in source (split change mutation)

fixing hang behavior if two barriers comes in the first two polls: wrap stream reader in mutex, we may run a bench later

- Summarize your change (**mandatory**)
- How does this PR work? Need a brief introduction for the changed logic (optional)
- Describe clearly one logical change and avoid lazy messages (optional)
- Describe any limitations of the current code (optional)
- Add the 'user-facing changes' label if your PR contains changes that are visible to users (optional)

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)

close #1285

### CODE REVIEW

**Comment by skyzh** [comment]
Please write all tests in https://github.com/singularity-data/risingwave/blob/main/src/compute/tests/integration_tests.rs, otherwise linker will OOM

**Comment by skyzh** [comment]
Please write all tests in https://github.com/singularity-data/risingwave/blob/main/src/compute/tests/integration_tests.rs, otherwise linker will OOM

**Comment by tabVersion** [comment]
> Please write all tests in https://github.com/singularity-data/risingwave/blob/main/src/compute/tests/integration_tests.rs, otherwise linker will OOM 

will fix

btw the test relies on kafka, how do I configure this when running test

**Comment by skyzh** [comment]
> btw the test relies on kafka, how do I configure this when running test

Unit test should be self-contained and should not depend on external service. Otherwise no developer will be willing to test their on changes with `./risedev test`.

**Comment by tabVersion** [comment]
Right. I will implement a source for testing

**Comment by tabVersion** [comment]
I will migrate to datagen connector after merge #3124

File: src/stream/src/executor/source.rs
Comment (tabVersion) [inline src/stream/src/executor/source.rs:823]: please just ignore the naming issue. I will write assert later.

**Review by shanicky — APPROVED** [review]
LGTM, we may need to test the performance of the mutex in `source.rs` later

**Comment by skyzh** [comment]
It's high probability that this impl will cause deadlock, I guess.

**Comment by skyzh** [comment]
Please hold this PR a little bit, I need to think for a while...

**Review by skyzh — APPROVED** [review]
LGTM. I have no clear idea on how to make it better. We may just fix it when there's really a deadlock, performance regression, or lose of message. The whole implementation looks counter-intuitive, but there might not be a better solution. Some thoughts below, but might not be applicable.

---

(1) Would be better if being implemented in the following way:

stream_reader is `Arc<parking_lot::Mutex<Option<Box<Reader>>>`

Before calling `.next`, we take the option out of the Mutex. After calling next, we put the reader back to Mutex. If mutex is not None, we can use the new reader for the next time.

(2) However, this approach still has problems. How can we ensure the reader is replaced to the new one after we finish process the barrier?

(3) The state machine looks pretty counter-intuitive. I think we should do a rewrite to make everything works. The basic logic should be like:

```
loop {
  let mut stream = select_with_strategy(stream, barrier);
  #[for_await]
  for item in stream {
    match item {
      Either::Right(normal_barrier) => ...
      Either::Right(conf_change_barrier) => break conf_change_barrier;
      Either::Left(chunk) => yield chunk
    }
  }
  if is_conf_change_barrier {
    change reader;
  }
}
```

(4) When there's pause and resume barrier in the future, how can we ensure we don't lose data in reader's stream state?

**Pull request merged**
